## Import Needed libraries

In [2]:
import pandas as pd

## Read in the attendance CSV to a data frame so you can do statistics on it

In [3]:
df = pd.read_csv('attendance.csv')

## Set up for questions

In [4]:
df.dtypes

Date          object
Away          object
Home          object
Attendance     int64
Type          object
dtype: object

We see that `Attendance` is an `int64`

## Question 1: Convert a numeric column to a smaller type

In [4]:
df.Attendance.nbytes

18960

In [7]:
df.Attendance.astype('Int32').nbytes

11850

## Question 2: Calculate the memory savings by converting to smaller numeric types

In [8]:
df.Attendance.nbytes - df.Attendance.astype('Int32').nbytes

7110

Converting to `Int32` we save 7110 bytes (or 7.11 kB)

## Question 3: Convert a string column into a categorical type

In [12]:
df.Away.astype('category')

0             Houston Astros
1       Arizona Diamondbacks
2       San Francisco Giants
3             Atlanta Braves
4             Tampa Bay Rays
                ...         
2365        Colorado Rockies
2366       Oakland Athletics
2367         Minnesota Twins
2368       Milwaukee Brewers
2369          Boston Red Sox
Name: Away, Length: 2370, dtype: category
Categories (30, object): ['Arizona Diamondbacks', 'Atlanta Braves', 'Baltimore Orioles', 'Boston Red Sox', ..., 'Tampa Bay Rays', 'Texas Rangers', 'Toronto Blue Jays', 'Washington Nationals']

## Question 4: Calculate the memory savings by converting to a categorical type

In [14]:
df.Away.astype('category').nbytes

2610

In [15]:
df.Away.nbytes

18960

In [16]:
df.Away.nbytes - df.Away.astype('category').nbytes

16350

We get a savings of 16.35 kB